servicetalk-client-api/src/main/java/io/servicetalk/client/api/BiTransportObserver.java

```diff
@@ -0,0 +1,351 @@
+/*
+ * Copyright © 2020 Apple Inc. and the ServiceTalk project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.servicetalk.client.api;
+
+import io.servicetalk.transport.api.ConnectionInfo;
+import io.servicetalk.transport.api.ConnectionObserver;
+import io.servicetalk.transport.api.ConnectionObserver.DataObserver;
+import io.servicetalk.transport.api.ConnectionObserver.MultiplexedObserver;
+import io.servicetalk.transport.api.ConnectionObserver.ReadObserver;
+import io.servicetalk.transport.api.ConnectionObserver.SecurityHandshakeObserver;
+import io.servicetalk.transport.api.ConnectionObserver.StreamObserver;
+import io.servicetalk.transport.api.ConnectionObserver.WriteObserver;
+import io.servicetalk.transport.api.TransportObserver;
+
+import javax.net.ssl.SSLSession;
+
+/**
+ * Combines two {@link TransportObserver}s into a single {@link TransportObserver}.
+ */
+final class BiTransportObserver implements TransportObserver {
+
+    private final TransportObserver first;
+    private final TransportObserver second;
+
+    /**
+     * Creates a new instance.
+     *
+     * @param first the {@link TransportObserver} that will receive events first
+     * @param second the {@link TransportObserver} that will receive events second
+     */
+    BiTransportObserver(final TransportObserver first, final TransportObserver second) {
+        this.first = first;
+        this.second = second;
+    }
+
+    @Override
+    public ConnectionObserver onNewConnection() {
+        return new BiConnectionObserver(first.onNewConnection(), second.onNewConnection());
+    }
+
+    private static final class BiConnectionObserver implements ConnectionObserver {
+
+        private final ConnectionObserver first;
+        private final ConnectionObserver second;
+
+        private BiConnectionObserver(final ConnectionObserver first, final ConnectionObserver second) {
+            this.first = first;
+            this.second = second;
+        }
+
+        @Override
+        public void onDataRead(final int size) {
+            try {
+                first.onDataRead(size);
+            } finally {
+                second.onDataRead(size);
```

Comment:
If `first` and `second` both throw then because `second` is called from a `finally` block, the first exception is lost. An alternative is to add separate `try-catch` blocks for each observer which also extends generically to `n` observers.

Reply:
I wanted to avoid safety guarantees for this implementation. Do you mind if I remove all try-catch blocks and will assume that first/second observers are never throw?
They can be wrapped by `CatchAllTransportObserver` impl as we discussed before.

Reply:
ok that sounds fair